src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,509 @@
-	m_active_clock_hi = _active_clock_hi;
-	m_last_clock_state = 0;
+inline bool cvsd_device_base::current_clock_state()
+{
+	// keep track of the clock state given its previous state and the number of samples produced
+	// i.e. if we generated m_samples_generated samples, at a sample rate of SAMPLE_RATE, then are we on a positive or negative level of a squarewave at clock() hz? SAMPLE_RATE may not be an integer multiple of clock()
+	//uint64_t fractions_of_second = (((uint64_t)m_samples_generated)<<32) / SAMPLE_RATE; // 32.32 bits of seconds passed so far
+	//uint32_t clock_edges_passed =  (fractions_of_second * clock() * 2)>>32
+	//return (((((uint64_t)m_samples_generated<<32) * clock() * 2 / SAMPLE_RATE)>>32) & 0x1)?true:false;
+	return (((uint64_t)m_samples_generated * clock() * 2 / SAMPLE_RATE) & 0x01)?true:false;
+}
 
-	/* create the stream */
-	m_channel = stream_alloc(0, 1, SAMPLE_RATE);
+void cvsd_device_base::digit_w(int digit)
+{
+	m_stream->update();
+	m_buffered_bit = digit ? true : false;
+}
 
-	save_item(NAME(m_last_clock_state));
-	save_item(NAME(m_digit));
-	save_item(NAME(m_new_digit));
-	save_item(NAME(m_shiftreg));
-	save_item(NAME(m_curr_sample));
-	save_item(NAME(m_next_sample));
-	save_item(NAME(m_update_count));
-	save_item(NAME(m_filter));
-	save_item(NAME(m_integrator));
+void cvsd_device_base::clock_w(int state)
+{
+	/* update the output buffer first */
+	m_stream->update();
+	bool clock_state = state ? true : false;
+
+	/* only makes sense for setups with a software driven clock */
+	assert(!is_external_oscillator());
+
+	/* speech clock changing? */
+	if (is_clock_changed(clock_state))
+	{
+		/* clear the update count */
+		m_samples_generated = 0;
+		process_bit(m_buffered_bit, clock_state);
+	}
+
+	/* update the clock */
+	m_last_clock_state = clock_state;
 }
 
-inline int hc55516_device::is_external_oscillator()
+int cvsd_device_base::clock_state_r()
 {
-	return clock() != 0;
+	/* only makes sense for setups with an external oscillator */
+	assert(is_external_oscillator());
+
+	m_stream->update();
+
+	return current_clock_state();
 }
 
+void cvsd_device_base::process_bit(bool bit, bool clock_state)
+{
+	// stub
+}
 
-inline int hc55516_device::is_active_clock_transition(int clock_state)
+//-------------------------------------------------
+//  sound_stream_update - handle a stream update
+//-------------------------------------------------
+
+void cvsd_device_base::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
 {
-	return (( m_active_clock_hi && !m_last_clock_state &&  clock_state) ||
-			(!m_active_clock_hi &&  m_last_clock_state && !clock_state));
+	// Stub, just return silence
+	auto &buffer = outputs[0];
+
+	m_samples_generated += buffer.samples();
+	if (m_samples_generated >= SAMPLE_RATE)
+		m_samples_generated -= SAMPLE_RATE;
+	for (int i = 0; i < buffer.samples(); i++)
+		buffer.put(i, stream_buffer::sample_t(0.0));
```

Comment:
I think there’s a helper to fill these with a fixed value called `fill` or something (I’d have to check the header to confirm).

Reply:
done, i switched this to the .fill(0); function